File: src/scpn_phase_orchestrator/nn/spectral.py
# SPDX-License-Identifier: AGPL-3.0-or-later | Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN Phase Orchestrator — Spectral analysis of coupling matrices

"""Differentiable spectral metrics for coupling matrix analysis.

All functions are differentiable via jnp.linalg.eigh, enabling
gradient-based topology optimisation: find the sparsest K that
maintains synchronisability above a target threshold.

Requires: jax>=0.4
"""

from __future__ import annotations

import jax
import jax.numpy as jnp


def laplacian_spectrum(K: jax.Array) -> jax.Array:
    """Sorted eigenvalues of the graph Laplacian L = D - K.

    Args:
        K: (N, N) symmetric coupling matrix (non-negative weights)

    Returns:
        (N,) eigenvalues in ascending order. First is ~0 (connected graph).
    """
    D = jnp.diag(jnp.sum(K, axis=1))
    L = D - K
    return jnp.linalg.eigh(L)[0]


def algebraic_connectivity(K: jax.Array) -> jax.Array:
    """Second-smallest Laplacian eigenvalue (Fiedler value).

    Measures how well-connected the network is. Zero iff disconnected.
    Differentiable — gradient flows through eigh.

    Args:
        K: (N, N) symmetric coupling matrix

    Returns:
        Scalar lambda_2
    """
    return laplacian_spectrum(K)[1]


def eigenratio(K: jax.Array) -> jax.Array:
    """Ratio lambda_N / lambda_2 (synchronisability metric).

    Lower eigenratio = more synchronisable (Barahona & Pecora 2002).
    The MSF (master stability function) approach shows that coupled
    oscillators synchronise when all transverse eigenvalues fall
    within the MSF stability interval.

    Args:
        K: (N, N) symmetric coupling matrix

    Returns:
        Scalar lambda_N / lambda_2
    """
    eigs = laplacian_spectrum(K)
    lambda_2 = eigs[1]
    lambda_N = eigs[-1]
    return lambda_N / jnp.maximum(lambda_2, 1e-10)


def sync_threshold(
    K: jax.Array,
    omegas: jax.Array,
) -> jax.Array:
    """Critical coupling strength estimate (Dorfler & Bullo 2014).

    K_c ≈ max|ω_i - ω_j| / lambda_2

    Below K_c, the network cannot synchronise. Above, it can.

    Args:
        K: (N, N) symmetric coupling matrix
        omegas: (N,) natural frequencies

    Returns:
        Scalar estimated critical coupling
    """
    lambda_2 = algebraic_connectivity(K)
    omega_spread = jnp.max(omegas) - jnp.min(omegas)
    return omega_spread / jnp.maximum(lambda_2, 1e-10)

File: julia/spectral.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN Phase Orchestrator — Spectral graph eigendecomposition (Julia port)

"""
spectral.jl — symmetric eigendecomposition of the combinatorial
graph Laplacian ``L = D − A`` where ``A`` is the reciprocal
undirected magnitude adjacency ``(|W| + |W'|) / 2`` with zeroed
diagonal.

``spectral_eig(knm_flat, n) -> (eigvals::Vector{Float64},
fiedler::Vector{Float64})``

Returns the eigenvalues in ascending order and the eigenvector
corresponding to ``λ₂`` (index 1 in Julia's 1-based indexing →
index 2 of the sorted array, which is the *second* smallest).
Uses Julia's ``LinearAlgebra.eigen`` which dispatches to LAPACK
``dsyev`` underneath — same numerics as NumPy and Mojo FFI.
"""

module SpectralJL

using LinearAlgebra

export spectral_eig


function spectral_eig(knm_flat::AbstractVector{Float64}, n::Integer)
    length(knm_flat) == n * n || error("knm_flat shape mismatch")
    W = reshape(abs.(knm_flat), (n, n))
    for i in 1:n
        W[i, i] = 0.0
    end
    A = 0.5 * (W + W')
    for i in 1:n
        A[i, i] = 0.0
    end
    degrees = vec(sum(A, dims=2))
    L = Diagonal(degrees) - A
    F = eigen(Symmetric(L))
    eigvals = Vector{Float64}(F.values)
    # Second smallest eigenvector (columns are eigenvectors,
    # Julia is 1-based, so column index 2).
    fiedler = Vector{Float64}(F.vectors[:, 2])
    return eigvals, fiedler
end

end  # module
